plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -1,915 +1,1218 @@
+		cpuMetrics:     p.CPUMetrics,
+		packageMetrics: p.PackageMetrics,
+		includedCPUs:   p.parsedIncludedCores,
+		excludedCPUs:   p.parsedExcludedCores,
+		perfEventFile:  p.EventDefinitions,
+		msrReadTimeout: time.Duration(p.MsrReadTimeout),
+		log:            p.Log,
+	})
 
-	if !p.areCoreMetricsEnabled() && !p.areGlobalMetricsEnabled() {
-		return fmt.Errorf("all configuration options are empty or invalid. Did not find anything to gather")
+	var err error
+	var initErr *ptel.MultiError
+	p.fetcher, err = ptel.New(opts...)
+	if err != nil {
+		if !errors.As(err, &initErr) {
+			return fmt.Errorf("failed to initialize metric fetcher interface: %w", err)
+		}
+		p.Log.Warnf("Plugin initialized with errors: %v", err)
 	}
 
-	p.fillCPUBusClock()
 	return nil
 }
 
-func (p *PowerStat) initMSR() {
-	// Initialize MSR service only when there is at least one metric enabled
-	if p.cpuFrequency || p.cpuBusyFrequency || p.cpuTemperature || p.cpuC0StateResidency || p.cpuC1StateResidency ||
-		p.cpuC6StateResidency || p.cpuBusyCycles || p.packageTurboLimit || p.packageUncoreFrequency || p.packageCPUBaseFrequency {
-		p.msr = newMsrServiceWithFs(p.Log, p.fs)
+// Stop deactivates perf events if one or more of the requested metrics rely on perf.
+func (p *PowerStat) Stop() {
+	if !p.needsPerf {
+		return
 	}
-}
 
-func (p *PowerStat) initRaplService() {
-	if p.packageCurrentPowerConsumption || p.packageCurrentDramPowerConsumption || p.packageThermalDesignPower || p.packageTurboLimit ||
-		p.packageUncoreFrequency || p.packageCPUBaseFrequency {
-		p.rapl = newRaplServiceWithFs(p.Log, p.fs)
+	if err := p.fetcher.DeactivatePerfEvents(); err != nil {
+		p.Log.Errorf("Failed to deactivate perf events: %v", err)
 	}
 }
 
-// fill CPUBusClockValue if required
-func (p *PowerStat) fillCPUBusClock() {
-	if p.packageCPUBaseFrequency {
-		// cpuBusClock is the same for every core/socket.
-		busClockInfo := p.getBusClock("0")
-		if busClockInfo == 0 {
-			p.Log.Warn("Disabling package metric: cpu_base_frequency_mhz. Can't detect bus clock value")
-			p.packageCPUBaseFrequency = false
-			return
-		}
+// Gather collects the plugin's metrics.
+func (p *PowerStat) Gather(acc telegraf.Accumulator) error {
+	// gather CPU metrics relying on coreFreq and msr which share CPU IDs.
+	if p.needsCoreFreq || p.needsMsrCPU {
+		p.addCPUMetrics(acc)
+	}
+
+	// gather CPU metrics relying on perf.
+	if p.needsPerf {
+		p.addCPUPerfMetrics(acc)
+	}
 
-		p.cpuBusClockValue = busClockInfo
+	// gather package metrics.
+	if len(p.PackageMetrics) != 0 {
+		p.addPackageMetrics(acc)
 	}
+
+	return nil
 }
 
-// Gather takes in an accumulator and adds the metrics that the Input gathers
-func (p *PowerStat) Gather(acc telegraf.Accumulator) error {
-	if p.areGlobalMetricsEnabled() {
-		p.addGlobalMetrics(acc)
+// parseConfig is a helper method that parses configuration fields from the receiver such as included/excluded CPU IDs.
+func (p *PowerStat) parseConfig() error {
+	var err error
```

Comment:
Can you please move this directly before use!?

Reply:
Absolutely. Done as part of commit [e2de6f6b](https://github.com/influxdata/telegraf/pull/14363/commits/e2de6f6b2577fe4e9ec7b420132b2750fe7e2654).